File: src/storage.rs
/// Storage and persistence layer
/// 
/// Handles checkpointing, output writing, compression, and metadata

use std::fs::File;
use std::io::{Write, BufWriter};
use std::path::Path;
use chrono::Local;

pub struct StorageWriter {
    output_path: std::path::PathBuf,
    compression: Option<String>,
    buffer_size: usize,
}

impl StorageWriter {
    pub fn new(output_path: impl AsRef<Path>, compression: Option<String>) -> Self {
        Self {
            output_path: output_path.as_ref().to_path_buf(),
            compression,
            buffer_size: 8192,
        }
    }

    /// Write tokens to file
    pub fn write_tokens(&self, tokens: &[String]) -> crate::Result<()> {
        // Create output directory if needed
        if let Some(parent) = self.output_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let file = File::create(&self.output_path)?;
        let mut writer = BufWriter::with_capacity(self.buffer_size, file);

        match self.compression.as_deref() {
            Some("gzip") => self.write_gzipped(&mut writer, tokens)?,
            Some("bzip2") => self.write_bzip2(&mut writer, tokens)?,
            Some("lz4") => self.write_lz4(&mut writer, tokens)?,
            Some("zstd") => self.write_zstd(&mut writer, tokens)?,
            None => self.write_plain(&mut writer, tokens)?,
            Some(fmt) => return Err(crate::Error::StorageError(
                format!("Unsupported compression: {}", fmt)
            )),
        }

        writer.flush()?;
        Ok(())
    }

    fn write_plain(&self, writer: &mut BufWriter<File>, tokens: &[String]) -> crate::Result<()> {
        for token in tokens {
            writeln!(writer, "{}", token)?;
        }
        Ok(())
    }

    fn write_gzipped(&self, writer: &mut BufWriter<File>, tokens: &[String]) -> crate::Result<()> {
        use flate2::Compression;
        use flate2::write::GzEncoder;

        let mut encoder = GzEncoder::new(writer, Compression::best());
        for token in tokens {
            writeln!(encoder, "{}", token)?;
        }
        encoder.finish()?;
        Ok(())
    }

    fn write_bzip2(&self, writer: &mut BufWriter<File>, tokens: &[String]) -> crate::Result<()> {
        use bzip2::write::BzEncoder;
        use bzip2::Compression;

        let mut encoder = BzEncoder::new(writer, Compression::best());
        for token in tokens {
            writeln!(encoder, "{}", token)?;
        }
        encoder.finish()?;
        Ok(())
    }

    fn write_lz4(&self, writer: &mut BufWriter<File>, tokens: &[String]) -> crate::Result<()> {
        use lz4_flex::frame::FrameEncoder;

        let mut encoder = FrameEncoder::new(writer);
        for token in tokens {
            writeln!(encoder, "{}", token)?;
        }
        encoder.finish()?;
        Ok(())
    }

    fn write_zstd(&self, writer: &mut BufWriter<File>, tokens: &[String]) -> crate::Result<()> {
        use zstd::stream::write::Encoder;

        let mut encoder = Encoder::new(writer, 21)?;
        for token in tokens {
            writeln!(encoder, "{}", token)?;
        }
        encoder.finish()?;
        Ok(())
    }

    /// Write JSONL format
    pub fn write_jsonl(&self, tokens: &[String]) -> crate::Result<()> {
        if let Some(parent) = self.output_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let file = File::create(&self.output_path)?;
        let mut writer = BufWriter::with_capacity(self.buffer_size, file);

        for token in tokens {
            let json = serde_json::json!({ "token": token });
            writeln!(writer, "{}", json)?;
        }

        writer.flush()?;
        Ok(())
    }

    /// Write CSV format
    pub fn write_csv(&self, tokens: &[String]) -> crate::Result<()> {
        if let Some(parent) = self.output_path.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let file = File::create(&self.output_path)?;
        let mut writer = BufWriter::with_capacity(self.buffer_size, file);

        writeln!(writer, "token,length,entropy")?;
        
        for token in tokens {
            let entropy = crate::filters::calculate_entropy(token);
            writeln!(writer, "\"{}\",{},{:.2}", token, token.len(), entropy)?;
        }

        writer.flush()?;
        Ok(())
    }
}

/// Checkpoint manager for resumable generation
pub struct CheckpointManager {
    checkpoint_dir: std::path::PathBuf,
}

impl CheckpointManager {
    pub fn new(checkpoint_dir: impl AsRef<Path>) -> crate::Result<Self> {
        let dir = checkpoint_dir.as_ref();
        std::fs::create_dir_all(dir)?;
        Ok(Self {
            checkpoint_dir: dir.to_path_buf(),
        })
    }

    /// Save checkpoint
    pub fn save_checkpoint(
        &self,
        job_id: &str,
        state: &CheckpointState,
    ) -> crate::Result<()> {
        let path = self.checkpoint_dir.join(format!("{}.json", job_id));
        let json = serde_json::to_string_pretty(state)?;
        std::fs::write(path, json)?;
        Ok(())
    }

    /// Load checkpoint
    pub fn load_checkpoint(&self, job_id: &str) -> crate::Result<Option<CheckpointState>> {
        let path = self.checkpoint_dir.join(format!("{}.json", job_id));
        if !path.exists() {
            return Ok(None);
        }

        let content = std::fs::read_to_string(&path)?;
        let state = serde_json::from_str(&content)?;
        Ok(Some(state))
    }

    /// List all checkpoints
    pub fn list_checkpoints(&self) -> crate::Result<Vec<String>> {
        let mut checkpoints = Vec::new();
        for entry in std::fs::read_dir(&self.checkpoint_dir)? {
            let entry = entry?;
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) == Some("json") {
                if let Some(name) = path.file_stem().and_then(|s| s.to_str()) {
                    checkpoints.push(name.to_string());
                }
            }
        }
        Ok(checkpoints)
    }

    /// Delete checkpoint
    pub fn delete_checkpoint(&self, job_id: &str) -> crate::Result<()> {
        let path = self.checkpoint_dir.join(format!("{}.json", job_id));
        if path.exists() {
            std::fs::remove_file(path)?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct CheckpointState {
    pub job_id: String,
    pub timestamp: String,
    pub config: crate::Config,
    pub last_token: Option<String>,
    pub tokens_generated: u64,
    pub current_length: usize,
    pub start_index: usize,
}

impl CheckpointState {
    pub fn new(job_id: String, config: crate::Config) -> Self {
        Self {
            job_id,
            timestamp: Local::now().to_rfc3339(),
            config,
            last_token: None,
            tokens_generated: 0,
            current_length: 0,
            start_index: 0,
        }
    }

    pub fn update(&mut self, token: &str, index: usize) {
        self.last_token = Some(token.to_string());
        self.tokens_generated += 1;
        self.start_index = index;
        self.timestamp = Local::now().to_rfc3339();
    }
}

/// Job metadata
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct JobMetadata {
    pub job_id: String,
    pub created_at: String,
    pub config: crate::Config,
    pub status: JobStatus,
    pub tokens_count: u64,
    pub output_file: Option<std::path::PathBuf>,
    pub estimated_cardinality: u64,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub enum JobStatus {
    Pending,
    Running,
    Completed,
    Failed(String),
    Paused,
}

pub struct JobManager {
    jobs_dir: std::path::PathBuf,
}

impl JobManager {
    pub fn new(jobs_dir: impl AsRef<Path>) -> crate::Result<Self> {
        let dir = jobs_dir.as_ref();
        std::fs::create_dir_all(dir)?;
        Ok(Self {
            jobs_dir: dir.to_path_buf(),
        })
    }

    /// Save job metadata
    pub fn save_job(&self, job: &JobMetadata) -> crate::Result<()> {
        let path = self.jobs_dir.join(format!("{}.json", job.job_id));
        let json = serde_json::to_string_pretty(job)?;
        std::fs::write(path, json)?;
        Ok(())
    }

    /// Load job metadata
    pub fn load_job(&self, job_id: &str) -> crate::Result<Option<JobMetadata>> {
        let path = self.jobs_dir.join(format!("{}.json", job_id));
        if !path.exists() {
            return Ok(None);
        }

        let content = std::fs::read_to_string(&path)?;
        let job = serde_json::from_str(&content)?;
        Ok(Some(job))
    }

    /// List all jobs
    pub fn list_jobs(&self) -> crate::Result<Vec<JobMetadata>> {
        let mut jobs = Vec::new();
        for entry in std::fs::read_dir(&self.jobs_dir)? {
            let entry = entry?;
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) == Some("json") {
                if let Ok(content) = std::fs::read_to_string(&path) {
                    if let Ok(job) = serde_json::from_str(&content) {
                        jobs.push(job);
                    }
                }
            }
        }
        Ok(jobs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_storage_writer() -> crate::Result<()> {
        let temp_dir = TempDir::new()?;
        let output_path = temp_dir.path().join("test.txt");
        let writer = StorageWriter::new(&output_path, None);
        
        let tokens = vec!["hello".to_string(), "world".to_string()];
        writer.write_tokens(&tokens)?;
        
        let content = std::fs::read_to_string(&output_path)?;
        assert!(content.contains("hello"));
        assert!(content.contains("world"));
        
        Ok(())
    }

    #[test]
    fn test_checkpoint_manager() -> crate::Result<()> {
        let temp_dir = TempDir::new()?;
        let manager = CheckpointManager::new(temp_dir.path())?;
        
        let mut state = CheckpointState::new("test_job".to_string(), crate::Config::default());
        state.update("token1", 0);
        
        manager.save_checkpoint("test_job", &state)?;
        let loaded = manager.load_checkpoint("test_job")?;
        
        assert!(loaded.is_some());
        let loaded_state = loaded.unwrap();
        assert_eq!(loaded_state.job_id, "test_job");
        
        Ok(())
    }
}

File: omniwordlist/storage.py
"""
Storage and persistence layer

Handles checkpointing, output writing, compression, and metadata
"""

import gzip
import bz2
import json
from pathlib import Path
from typing import Iterator, Optional
from .error import StorageError


class OutputWriter:
    """Base output writer"""
    
    def __init__(self, path: Path, compression: Optional[str] = None, format: str = "txt"):
        """
        Initialize output writer
        
        Args:
            path: Output file path
            compression: Compression format (gzip, bzip2, lz4, zstd)
            format: Output format (txt, jsonl, csv)
        """
        self.path = path
        self.compression = compression
        self.format = format
        self.file_handle = None
        self.bytes_written = 0
        self.lines_written = 0
    
    def open(self):
        """Open output file"""
        # Ensure parent directory exists
        self.path.parent.mkdir(parents=True, exist_ok=True)
        
        # Open with appropriate compression
        if self.compression == "gzip":
            self.file_handle = gzip.open(self.path, 'wt', encoding='utf-8')
        elif self.compression == "bzip2":
            self.file_handle = bz2.open(self.path, 'wt', encoding='utf-8')
        elif self.compression == "lz4":
            try:
                import lz4.frame
                self.file_handle = lz4.frame.open(self.path, 'wt', encoding='utf-8')
            except ImportError:
                raise StorageError("lz4 compression requires lz4 package")
        elif self.compression == "zstd":
            try:
                import zstandard as zstd
                cctx = zstd.ZstdCompressor()
                self.file_handle = cctx.stream_writer(open(self.path, 'wb'))
            except ImportError:
                raise StorageError("zstd compression requires zstandard package")
        else:
            self.file_handle = open(self.path, 'w', encoding='utf-8')
        
        # Write CSV header if needed
        if self.format == "csv":
            self._write_line("token,entropy,length")
    
    def write(self, token: str, metadata: dict = None):
        """
        Write a token to output
        
        Args:
            token: Token to write
            metadata: Optional metadata
        """
        if not self.file_handle:
            raise StorageError("Output file not opened")
        
        if self.format == "txt":
            line = token + "\n"
        elif self.format == "jsonl":
            from .filters import calculate_entropy
            data = {
                "token": token,
                "entropy": calculate_entropy(token),
                "length": len(token)
            }
            if metadata:
                data.update(metadata)
            line = json.dumps(data) + "\n"
        elif self.format == "csv":
            from .filters import calculate_entropy
            entropy = calculate_entropy(token)
            line = f'"{token}",{entropy},{len(token)}\n'
        else:
            line = token + "\n"
        
        self._write_line(line)
    
    def _write_line(self, line: str):
        """Internal method to write line"""
        if self.compression == "zstd":
            # zstd needs bytes
            self.file_handle.write(line.encode('utf-8'))
        else:
            self.file_handle.write(line)
        
        self.bytes_written += len(line.encode('utf-8'))
        self.lines_written += 1
    
    def close(self):
        """Close output file"""
        if self.file_handle:
            if self.compression == "zstd":
                # Flush zstd compressor
                try:
                    self.file_handle.flush()
                except:
                    pass
            self.file_handle.close()
            self.file_handle = None
    
    def __enter__(self):
        """Context manager entry"""
        self.open()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        self.close()


class CheckpointManager:
    """Manage generation checkpoints for resume capability"""
    
    def __init__(self, checkpoint_dir: Path):
        """
        Initialize checkpoint manager
        
        Args:
            checkpoint_dir: Directory for checkpoint files
        """
        self.checkpoint_dir = checkpoint_dir
        self.checkpoint_dir.mkdir(parents=True, exist_ok=True)
    
    def save_checkpoint(self, job_id: str, state: dict):
        """
        Save checkpoint state
        
        Args:
            job_id: Job identifier
            state: State dictionary to save
        """
        checkpoint_path = self.checkpoint_dir / f"{job_id}.checkpoint.json"
        with open(checkpoint_path, 'w') as f:
            json.dump(state, f, indent=2)
    
    def load_checkpoint(self, job_id: str) -> Optional[dict]:
        """
        Load checkpoint state
        
        Args:
            job_id: Job identifier
            
        Returns:
            State dictionary or None
        """
        checkpoint_path = self.checkpoint_dir / f"{job_id}.checkpoint.json"
        if not checkpoint_path.exists():
            return None
        
        with open(checkpoint_path, 'r') as f:
            return json.load(f)
    
    def delete_checkpoint(self, job_id: str):
        """
        Delete checkpoint
        
        Args:
            job_id: Job identifier
        """
        checkpoint_path = self.checkpoint_dir / f"{job_id}.checkpoint.json"
        if checkpoint_path.exists():
            checkpoint_path.unlink()


def write_tokens_to_file(tokens: Iterator[str], output_path: Path, 
                        compression: Optional[str] = None, 
                        format: str = "txt") -> int:
    """
    Write tokens to file with optional compression
    
    Args:
        tokens: Iterator of tokens
        output_path: Output file path
        compression: Optional compression format
        format: Output format
        
    Returns:
        Number of tokens written
    """
    count = 0
    with OutputWriter(output_path, compression, format) as writer:
        for token in tokens:
            writer.write(token)
            count += 1
    return count
